src/vs/workbench/parts/terminal/electron-browser/terminalInstance.ts

```diff
@@ -846,6 +867,57 @@ export class TerminalInstance implements ITerminalInstance {
 			this._messageTitleListener = null;
 		}
 	}
+
+	private static executeWMIC(pid: number): TPromise<{ executable: string, pid: number }[]> {
+		return new TPromise((resolve, reject) => {
+			cp.execFile('wmic.exe', ['process', 'where', `parentProcessId=${pid}`, 'get', 'ExecutablePath,ProcessId'], (err, stdout, stderr) => {
+				if (err) {
+					reject(err);
+				} else if (stderr.length > 0) {
+					resolve([]); // No processes found
+				} else {
+					resolve(stdout.split('\n').slice(1).filter(str => !/^\s*$/.test(str)).map(str => {
+						const s = str.split('  ');
+						return { executable: s[0], pid: Number(s[1]) };
+					}));
+				}
+			});
+		});
+	}
+
+	private getChildProcesses(pid: number): TPromise<string> {
+		return TerminalInstance.executeWMIC(pid).then(result => {
+			if (result.length === 0) {
+				if (this._pidStack.length > 1) {
+					this._pidStack.pop();
+					return this.getChildProcesses(this._pidStack[this._pidStack.length - 1]);
+				}
+				return TPromise.as([]);
+			}
+			this._pidStack.push(result[0].pid);
+			return TPromise.as(result[0].executable);
+		}, error => { return error; });
+	}
+
+	public getShellName(): TPromise<string> {
+		if (platform.platform !== platform.Platform.Windows) {
+			return TPromise.as(null);
+		}
+		if (this._pidStack.length === 0) {
+			this._pidStack.push(this._processId);
+		}
+		return new TPromise<string>((resolve) => {
+			// wait 100ms before running getChildProcesses
+			setTimeout(() => {
+				this.getChildProcesses(this._pidStack[this._pidStack.length - 1]).then(result => {
+					if (result.length > 0) {
+						resolve(result);
+					}
+					resolve(this._shellLaunchConfig.executable);
+				}, error => { return error; });
+			}, 100);
```

Comment:
We should move this to a named constant (especially because the comment explaining is so far from this line.